odl/discr/grid.py

```diff
@@ -689,80 +689,81 @@ def meshgrid(self):
         >>> g[0, ..., 1:]
         TensorGrid([-1.0], [2.0, 4.0], [5.0], [4.0, 7.0])
         >>> g[::2, ..., ::2]
         TensorGrid([-1.0, 3.0], [2.0, 4.0], [5.0], [2.0, 7.0])
-        """
-        slc_list = list(np.atleast_1d(np.s_[slc]))
-        if None in slc_list:
-            raise IndexError('Creation of new axes not supported.')
-
-        try:
-            # All single indices
-            idx = np.array(slc_list).astype('int64', casting='safe')
-            if len(idx) < self.ndim:
-                raise IndexError('too few indices ({} < {}).'
-                                 ''.format(len(idx), self.ndim))
-            elif len(idx) > self.ndim:
-                raise IndexError('too many indices ({} > {}).'
-                                 ''.format(len(idx), self.ndim))
 
-            return np.array([v[i] for i, v in zip(idx, self.coord_vectors)])
-
-        except TypeError:
-            pass
-
-        if Ellipsis in slc_list:
-            if slc_list.count(Ellipsis) > 1:
-                raise IndexError("Cannot use more than one ellipsis.")
-            if len(slc_list) == self.ndim + 1:  # Ellipsis without effect
-                ellipsis_idx = self.ndim
-                num_after_ellipsis = 0
-            else:
-                ellipsis_idx = slc_list.index(Ellipsis)
-                num_after_ellipsis = len(slc_list) - ellipsis_idx - 1
-            slc_list.remove(Ellipsis)
+        Too few indices are filled up with an ellipsis from the right:
 
+        >>> g[0]
+        TensorGrid([-1.0], [2.0, 4.0], [5.0], [2.0, 4.0, 7.0])
+        >>> g[0] == g[0, :, :, :] == g[0, ...]
+        True
+        """
+        # Support indexing with fewer indices as indexing along the first
+        # corresponding axes. In the other cases, normalize the input.
+        if np.isscalar(indices):
+            indices = [indices, Ellipsis]
+        elif (isinstance(indices, slice) or
+              (not isinstance(indices, np.ndarray) and indices == Ellipsis)):
+            indices = [indices]
+
+        indices = list(indices)
+        if len(indices) < self.ndim and Ellipsis not in indices:
+            indices.append(Ellipsis)
+
+        # Turn Ellipsis into the correct number of slice(None)
+        if Ellipsis in indices:
+            if indices.count(Ellipsis) > 1:
+                raise ValueError('cannot use more than one Ellipsis.')
+
+            eidx = indices.index(Ellipsis)
+            extra_dims = self.ndim - len(indices) + 1
+            indices = (indices[:eidx] + [slice(None)] * extra_dims +
+                       indices[eidx + 1:])
+
+        # Catch most common errors
+        if any(s.start == s.stop and s.start is not None or
+               s.start == n
+               for s, n in zip(indices, self.shape) if isinstance(s, slice)):
+            raise ValueError('Slices with empty axes not allowed.')
+        if None in indices:
+            raise ValueError('creating new axes is not supported.')
```

Comment:
'<b>C</b>reating new axes is not supported. <b>Use ``insert``.</b>'

Perhaps? Im not sure, we soldom do this (add tips) but this could be one of those places.